executor/executor.go

```diff
@@ -1732,6 +1732,11 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 		if variable.TopSQLEnabled() && prepareStmt.SQLDigest != nil {
 			topsql.AttachSQLInfo(goCtx, prepareStmt.NormalizedSQL, prepareStmt.SQLDigest, "", nil, vars.InRestrictedSQL)
 		}
+		if s, ok := prepareStmt.PreparedAst.Stmt.(*ast.SelectStmt); ok {
```

Comment:
We could use `isReadOnly` to check if the statment will write any data and allow just read-only statment to use it, this limination need to be handled very carefully othewise the data consistency could be broken.

Reply:
We can not rely on `IsReadOnly` from planner completely, in that function, like show stmt queries are also considered as read-only.

Reply:
Could we also add the `IsReadOnly` check in the `isWeakConsistencyRead` logic it could be safer.